datafusion/physical-plan/src/aggregates/group_values/group_column.rs

```diff
@@ -62,57 +64,60 @@ pub trait GroupColumn: Send + Sync {
 
 pub struct PrimitiveGroupValueBuilder<T: ArrowPrimitiveType> {
     group_values: Vec<T::Native>,
-    nulls: Vec<bool>,
-    // whether the array contains at least one null, for fast non-null path
-    has_null: bool,
-    nullable: bool,
+    /// Null state (when None, input is guaranteed not to have nulls)
+    nulls: Option<MaybeNullBufferBuilder>,
```

Comment:
I collapsed `nullable` and `has_nulls` into an `Option` and an `Enum` which I think makes the intent much clearer (and harder to mess up the `nullable` optimization

Reply:
My benchmark shows a slowdown compared to the main branch. I experienced a similar slowdown before due to the use of Option, and I suspect that might be the case again with this change

Reply:
My bencmarks seem to show a speedup, but I did not compare just a bool. I'll make that change and rerun